polkadot/xcm/xcm-executor/src/assets.rs

```diff
@@ -132,41 +132,20 @@ impl AssetsInHolding {
 
 	/// Mutate `self` to contain all given `assets`, saturating if necessary.
 	///
-	/// NOTE: [`AssetsInHolding`] are always sorted, allowing us to optimize this function from
-	/// `O(n^2)` to `O(n)`.
+	/// NOTE: [`AssetsInHolding`] are always sorted
 	pub fn subsume_assets(&mut self, mut assets: AssetsInHolding) {
-		let mut f_iter = assets.fungible.iter_mut();
-		let mut g_iter = self.fungible.iter_mut();
-		if let (Some(mut f), Some(mut g)) = (f_iter.next(), g_iter.next()) {
-			loop {
-				if f.0 == g.0 {
-					// keys are equal. in this case, we add `self`'s balance for the asset onto
-					// `assets`, balance, knowing that the `append` operation which follows will
-					// clobber `self`'s value and only use `assets`'s.
-					(*f.1).saturating_accrue(*g.1);
-
-					f = match f_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-					g = match g_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-				} else if f.0 < g.0 {
-					f = match f_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-				} else {
-					g = match g_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-				}
-			}
+		let assets_iter = assets.fungible.into_iter();
+		// for fungibles, find matching fungibles and sum their amounts so we end-up having just
+		// single such fungible but with increased amount inside
+		for (k, v) in assets_iter {
```

Comment:
Do we need to create this assets_iter?
```suggestion
		// for fungibles, find matching fungibles and sum their amounts so we end-up having just
		// single such fungible but with increased amount inside
		for (k, v) in assets.fungible {
```